hexkit/custom_types.py

```diff
@@ -31,13 +31,35 @@
 Ascii = str
 
 
-# A ContextConstructable is a class with a classmethod `construct` that creates an async
+# A AsyncConstructable is a class with a async classmethod `construct` that is used when
+# asynchronous constuction/instantiation logic is needed (which cannot be handeled in
+# a synchronous __init__ method).
+# With the current typing features of Python, it seems not possible to correctly type
+# a class with that signature.
+# E.g. one could define a typing.Protocol subclass with following method:
+# ```
+# class AsyncConstructable(Protocol):
+#     @classmethod
+#     def construct(cls, *args: Any, **kwargs: Any): ...
+# ```
+# However, this is incompatible with implementations that don't explicitly use `*args`
+# and `*kwargs`, e.g. the following method does not comply with the above function stub:
+# ```
+# class SomeImplementation:
+#     @classmethod
+#     def construct(cls, foo: str): ...
+# ```
+# Thus using a type alias for now:
+AsyncConstructable = Any
```

Comment:
This should work and is a little bit better:

```python
class AsyncConstructable(Protocol):
    """A class that can be asynchronously constructed.

    An AsyncConstructable is a class with an async classmethod `construct`
    that is used when asynchronous constuction/instantiation logic is needed
    (which cannot be handeled in a synchronous __init__ method).
    """

    construct: Callable[..., Awaitable["AsyncConstructable"]]
```

Reply:
I already tried this some weeks ago but mypy is not recognizing this.
Will push a proof in a minute.

Since we are already making checks for the correct type at runtime, I think there is little value in investing much more time into that right now. If you agree, I can put it on the refactoring list.

Reply:
Please have a look at this branch: https://github.com/ghga-de/hexkit/pull/36
Where mypy is reporting the following error: https://github.com/ghga-de/hexkit/actions/runs/3248759888/jobs/5330354785#step:6:169